django/utils/timesince.py

```diff
@@ -63,7 +63,11 @@ def timesince(d, now=None, reversed=False, time_strings=None, depth=2):
     if now and not isinstance(now, datetime.datetime):
         now = datetime.datetime(now.year, now.month, now.day)
 
-    now = now or datetime.datetime.now(datetime.timezone.utc if is_aware(d) else None)
+    if is_aware(d):
+        # Compared datetimes must be in the same time zone.
+        now = now.astimezone(d.tzinfo) if now else datetime.datetime.now(tz=d.tzinfo)
+    else:
+        now = now or datetime.datetime.now()
```

Comment:
I would have written the new logic a bit differently, something like:

```python
if not now:
    now = datetime.datetime.now()
if is_aware(d):
    now = now.astimezone(d.tzinfo)
```
to clearly separate concerns about settings a default for `now`, and on the other hand reaching time zone parity with `d`.

In any case the proposed changes looks good to me, thanks for working on this.

Reply:
This should return nothing when we have naive `now` and aware `d`, see:
- `template_tests.filter_tests.test_timesince.TimesinceTests.test_timesince16`
- `template_tests.filter_tests.test_timeuntil.TimeuntilTests.test_aware_naive_type_error`

so I have to make it a bit more complicated.

Reply:
I love tests, our safety net! Looks good